design/coroutines.md

```diff
@@ -0,0 +1,688 @@
+# Coroutines in Flow
+
+* [Introduction](#Introduction)
+* [Basic Types](#basic-types)
+* [Choose-When](#choose-when)
+    * [Execution in when-expressions](#execution-in-when-expressions)
+    * [Waiting in When-Blocks](#waiting-in-when-blocks)
+* [Generators](#generators)
+    * [Generators and ranges](#generators-and-ranges)
+    * [Eager vs Lazy Execution](#eager-vs-lazy-execution)
+    * [Generators vs Promise Streams](#generators-vs-promise-streams)
+* [Uncancellable](#uncancellable)
+* [Porting ACTOR's to C++ Coroutines](#porting-actors-to-c-coroutines)
+    * [Lifetime of Locals](#lifetime-of-locals)
+    * [Unnecessary Helper Actors](#unnecessary-helper-actors)
+    * [Replace Locals with Temporaries](#replace-locals-with-temporaries)
+    * [Don't Wait in Error-Handlers](#dont-wait-in-error-handlers)
+    * [Make Static Functions Class Members](#make-static-functions-class-members)
+    * [Initialization of Locals](#initialization-of-locals)
+
```

Comment:
I wonder if you may add a section explaining how determinism is preserved with co-routines, assuming you have already validated this. I guess the co-routine runtime has a scheduler that is deterministic?

Reply:
I added a new section `Coroutines vs ACTORs` that hopefully clarifies this. But fundamentally, coroutines don't change anything in terms of scheduling, we still use `Net2` or the simulator main loop. C++ coroutines only handle the suspension and resumption of coroutines.

So the right way of thinking about this is that C++ coroutines are a replacement for the actor compiler, not for flow.